packages/daemon/DESIGN.md

```diff
@@ -0,0 +1,215 @@
+capability bearing objects if it restarts.
+
+When Endo runs a confined program, it provides a Powers object that the
+confined program can use to **request** powers from the user.
+The Powers object is entangled with a corresponding Inbox.
+The user interacts with the Inbox to respond to requests from the Powers
+holder.
+
+Both the Inbox held by the user (host) and the Powers held by program (guest)
+refer to formulas with pet names, and have separate pet name spaces.
+Changing a pet name does not alter or delete the underlying formula, but a
+formula that is unreachable from the user's pet names may be garbage collected.
+
+- The program can use the Powers object to request an ephemeral value
+  from the user.
+  The Powers will not retain the resulting formula.
+
+  ```js
+  const file = E(powers).request('a file');
+  ```
+
+- The program can use the Powers object to request a durable value
+  and give it a pet name.
+  The Powers will memoize the value and retain the resulting formula.
+  Endo will honor subsequent requests for the same pet name from this Powers
+  object by reconstructing the value from its formula if necessary.
+
+  ```js
+  const file = E(powers).request('a file, 'fileName');
+  ```
+
+- The user can respond to pending requests by streaming requests
+  from their Inbox.
+  The user can resolve the request with the pet name of a formula they hold,
+  giving the guest the corresponding value.
+  The user can reject the request with a message.
+  The user can observe whether the request is still pending, which will be
+  useful for a graphical user interface.
+
+  ```js
+  for await (const request of iterateRef(E(inbox).requests())) {
+    request.resolve('userPetName');
+  }
+  ```
+
+- The user can create a new Powers object by calling the `makePowers`
+  method of their Inbox.
+  In a user interface, requests will be attributed to their pet name for the
+  Powers object.
+
+Every formula has a string identifier.
+If the behavior of the formula cannot be inferred from the identifier,
+Endo stores a corresponding JSON description in a key value store (tentatively
+just files) indexed on the identifier.
+The description instructs Endo in how to create the corresponding value.
+Formulas with dependencies list the identifiers of those dependencies in a
+`slots` property then refer elsewhere to those formulas by the index of the
+slot.
+This makes formulas transparent to a garbage collector.
+
+- A blob formula denotes a content address hash for bytes that Endo has stored.
+  Endo provides a Blob reference for that formula with methods for reading the
+  text, bytes, or byte stream from the blob.
+
+  Formula Identifier:
+
+  ```
+  readableSha512/$SHA512
+  ```
+
+  Endo stores the blob in the user's home directory, tentatively in a file with
+  the same name as the formula identifier.
+
+- A worker formula denotes the unique identifier for an Endo worker.
```

Comment:
A node or browser worker is ephemeral. For example, a node or browser worker never survives a crash. What about an Endo worker? Can a worker formula be dereferenced to a later incarnation of the "same" worker?

Reply:
If an Endo daemon restarts, it can reconstruct a value for any formula that is still retained starting from the user’s pet store, including any workers and confined programs, including any of the values they were previously granted and gave a durable name.